### ISSUE TITLE
Certificate_Store find_cert method returns pointer to store-owned X509_Certificate object

### ISSUE THREAD

**Issue by al45tair** [randombit/botan#471]
The `find_cert` method on `Certificate_Store` is currently defined as follows:

```c++
virtual const X509_Certificate*
         find_cert(const X509_DN& subject_dn, const std::vector<byte>& key_id) const = 0;
```

The documentation seems to imply that the original intention was to return `std::vector<X509_Certificate>` from these kinds of methods, but this method is returning a pointer. Unfortunately, looking at the implementation of `Certificate_Store_In_Memory`, it becomes clear that the code currently expects the *store* to own the pointer. This is problematic if you want to implement an external certificate store, e.g. based on the OS X keychain or on some external server, because in those cases the `X509_Certificate` object would have to be allocated on the fly and there is no way for the store to know when it should deallocate such an object.

**Comment by randombit** [comment]
This area of the code is in need of some help. As can be seen by there not being any implementation for system-level cert stores, or even just a basic sqlite3-backed certstore, it has not been driven by application requirements overmuch but is just kind of the minimum interface that is necessary for testing and simple applications. And, as you rightly point out, it's too simple to even be extended properly for such real world uses. :/

Part of the problem is Certificate_Store represents two different things - a way of doing lookups on certs, generally and also being something to pass around with the idea of "my trust roots can be looked up via this particular object". This confusion also affects the API - does any real application actually need more than one Certificate_Store? Certainly not when they are all in memory anyway, but still `x509_path_validate` consumes a `vector` of them.

Returning a `shared_ptr<X509_Certificate>` from `find_cert` would be a good start since that would accommodate either in-memory or DB-backed stores.

Amusingly? the main user of this class, path validation, makes a full copy of the `X509_Certificate` object immediately, while building the path. It could instead keep the `shared_ptr`s, and avoid the copy.

**Comment by al45tair** [comment]
OK, in that case I might take a look at this area and see if I can come up with a sensible patch (including new documentation). I’m looking to use Botan’s TLS implementation in a cross-platform app, and I’d like it to use the system certificate stores (whether the System keychain on OS X or the Certificate Stores on Windows) rather than shipping around its own trust roots.

**Comment by flanfly** [comment]
Any update on this?

**Comment by randombit** [comment]
I have not touched this issue. But I'm open to API changes to better interop with system/token PKI stores.

**Comment by al45tair** [comment]
FWIW, I started working on it, then changed tack somewhat and decided that I might just not use Botan for my project after all. (The final straw was when I discovered that the Botan TLS implementation makes blocking networking calls for OCSP/CRL fetching; I want my project to use asynchronous networking (ala libuv) and it can't block like that or it will block activity on other connections serviced by the same event loop, so in addition to fixing the certificate handling, I'd also have to alter the TLS implementation itself quite a bit.)

**Comment by randombit** [comment]
> The final straw was when I discovered that the Botan TLS implementation makes blocking networking calls for OCSP/CRL fetching

Once the TLS callbacks type PR is merged it will be possible to add OCSP and CRL fetch request functions which return a `future` and allowing the application to control all networking and allow fully async applications. Current behavior is not great.

**Comment by flanfly** [comment]
My idea is to use the already present dependency on SQlite to implement a persistent store for certificates and private keys. This store would subclass ``Certificate_Store`` and add methods for adding and removing certificates, as well as managing private keys. The private keys can be encrypted before writing them to disk. Also, the store would hold an optional revocation status of certificates and allow generating CRLs for them. This way Botan can be used to manage a small PKI.

I'd also change the return value of ``find_cert`` to ``shared_ptr`` as @randombit suggested above. The class would look like this:
```c++
class BOTAN_DLL Certificate_Store_On_Disk  : public Certificate_Store
   {
   public:
      explicit Certificate_Store_On_Disk(const std::string& path,
                                         const secure_vector<byte>& passwd);

      // Certificate handling
      virtual const std::shared_ptr<X509_Certificate>
         find_cert(const X509_DN& subject_dn, const std::vector<byte>& key_id) const;

      virtual const std::shared_ptr<X509_CRL>
         find_crl_for(const X509_Certificate& subject) const;

      virtual std::vector<X509_DN> all_subjects() const;

      bool insert_cert(const X509_Certificate& cert);
      bool remove_cert(const X509_Certificate& cert);

      // Private key handling
      std::shared_ptr<Private_Key> find_key(const X509_Certificate&) const;
      std::vector<std::shared_ptr<X509_Certificate>>
         find_certs_for_key(const Private_Key&) const;

      bool insert_key(const X509_Certificate&, const Private_Key&);
      bool remove_key(const Private_Key&);

      // Revocation
      bool revoke_cert(const X509_Certificate&,CRL_Code);
      bool temporary_revoke_cert(const X509_Certificate&,CRL_Code,const X509_Time& timeout);
      bool affirm(const X509_Certificate&);

      X509_CRL generate_crl() const;
   }
```
Internally, the class would maintain a list of ``Private_Key``/``std::vector<X509_Certificate>`` pairs with the ``Private_Key`` being optional. What do you think?

**Comment by randombit** [comment]
Overall what you propose looks good. Some questions and comments:

Instead of `path` to the sqlite3 database, call your current class `Certificate_Store_in_SQL` taking a `shared_ptr<SQL_Database>` (no sqlite dependency). Then `Certificate_Store_On_Disk`, which does depend on sqlite3, subclasses Certificate_Store_in_SQL and creates the `Sqlite3_Database` object. A similar pattern can be seen in `TLS::Session_Manager_SQL` and `TLS::Session_Manager_Sqlite`. This allows some other DB type (or a mock DB for testing) to be easily defined and the cert management logic to be reused.

Add an additional argument `table_prefix` to the constructor which an application can optionally provide to override the default string prefixed to all tables ("botan_x509_" or whatever). This allows the application to use the same database for other purposes (application data, multiple PKIs under different namespaces, ...) without fear of table name conflicts.

Something that is not clear to me in your proposal - does `Certificate_Store_On_Disk` load all of the certs at startup, and then ignore the database thereafter, or would `find_cert` search the DB using SQL and then just save a `smart_ptr` (or `weak_ptr` to allow collection of no-longer-used cert objects?) for certificates which it had found via a search. The second makes concurrent use of the db useful and reduces memory usage, and so seems much preferable to me.

To allow searching you would probably want to save alongside the certificate blob, the subject DN sent to a short canonical string format like LDAP's, the subject key id (if any), the fingerprint (`X509_Certificate::fingerprint("SHA-256")`), and the hash of the private key if known. The fingerprint can be used as a primary key since it is always available and should never duplicate.

Then save private keys as H(k),E(k) to allow fast mapping between the cert and key tables, and as a de-duplicator.

Is there any plan to add integrity checks to the DB? As an example of something I mean - upon initialization, create a RSA keypair and save an encrypted copy to the database. For each operation (say `insert_cert`) save both the cert as well as current_time,SIG(database_key, current_time || cert). `Certificate_Store_On_Disk` verifies that the signature is valid before loading. This is maybe out of scope and certainly complicates things, but you can easily see the damage someone with plain write access to the database file can do. I'm fine with punting this issue to "make sure your OS access controls are good on the DB file" if you are, but if so this issue should be clearly documented.

**Comment by flanfly** [comment]
Thanks for the review, I agree with you on using a generic database interface and to allow table prefixes.

My plan was to search the database each time `find_cert` is called, allocate a `shared_ptr` and return it. Simply because it is easier and less error prone to implement, but I'm ok with caching them.

While I agree that integrity checks for the certificates are a good idea I would prefer to do this at a later time.

**Comment by randombit** [comment]
> My plan was to search the database each time find_cert is called, allocate a shared_ptr and return it. Simply because it is easier and less error prone to implement, but I'm ok with caching them.

Either of these approaches is fine with me. A cache of parsed certs is purely an optimization (which may not even be required or useful, depending on the app), and could be added afterwards without any API changes.

**Comment by randombit** [comment]
With #631 merged I think this issue can be closed, since the original topic was regarding supporting system certificate stores, and returning owned vs shared ptrs. We should be good in that regard now. The issues re the handling for OCSP requests was taken up in #653.
